### PR TITLE
Added the RangeFinder class to represent a depth camera

### PR DESCRIPTION
**Pull request by astumpf**
Add the rangefinder class that reads in a depth camera node from URDF and translates it as a rangefinder node into webots proto.

### CODE REVIEW

File: urdf2webots/parserURDF.py [hunk 1]
```diff
@@ -949,6 +990,42 @@ def parseGazeboElement(element, parentLink, linkList):
...
+                    if hasElement(imageElement, 'height'):
+                        rangefinder.height = int(imageElement.getElementsByTagName('height')[0].firstChild.nodeValue)
+                if hasElement(cameraElement, 'clip'):
+                    clipElement = cameraElement.getElementsByTagName('clip')[0]
+                    if hasElement(clipElement, 'near'):
+                        rangefinder.near = float(clipElement.getElementsByTagName('near')[0].firstChild.nodeValue)
+            if hasElement(sensorElement, 'noise'):
+                noiseElement = sensorElement.getElementsByTagName('noise')[0]
+                if hasElement(noiseElement, 'stddev'):
+                    rangefinder.noise = float(noiseElement.getElementsByTagName('stddev')[0].firstChild.nodeValue)
```
Comment (BenjaminHug8) [inline urdf2webots/parserURDF.py:1014]:
```suggestion
                    rangefinder.noise = float(noiseElement.getElementsByTagName('stddev')[0].firstChild.nodeValue)
                    if rangefinder.maxRange:
                            rangefinder.noise /= rangefinder.maxRange
```

With this change, the `noise` element should be parsed after the `range` one.

File: urdf2webots/parserURDF.py [hunk 2]
```diff
@@ -949,6 +990,42 @@ def parseGazeboElement(element, parentLink, linkList):
...
+                    rangefinder.minRange = float(rangeElement.getElementsByTagName('min')[0].firstChild.nodeValue)
+                if hasElement(rangeElement, 'max'):
+                    rangefinder.maxRange = float(rangeElement.getElementsByTagName('max')[0].firstChild.nodeValue)
+                if hasElement(rangeElement, 'resolution'):
+                    rangefinder.resolution = float(rangeElement.getElementsByTagName('resolution')[0].firstChild.nodeValue)
+            # minRange and near default values are 0.01 in Webots; ensure constraint near <= minRange
+            if not rangefinder.near and rangefinder.minRange < 0.01:
+                rangefinder.near = rangefinder.minRange
+            if not rangefinder.minRange and rangefinder.near > 0.01:
+                rangefinder.minRange = rangefinder.near
```
Comment (BenjaminHug8) [inline urdf2webots/parserURDF.py:1027]: Same remark as for [here](https://github.com/cyberbotics/urdf2webots/pull/160/files#r811815242)

**Review by BenjaminHug8 — CHANGES_REQUESTED** [review]
Thank you once again for your contribution!

I let you check the requested changes.

**Comment by BenjaminHug8** [comment]
This PR should also target `develop` instead of `master`.

File: urdf2webots/parserURDF.py (hunk 2 above)
Comment (astumpf) [reply urdf2webots/parserURDF.py]: Changes applied as commented

**Comment by astumpf** [comment]
Still have some minor issues with the rangefinder, which I hope you can figure out quickly.

First, the resulting proto:
 
```
RangeFinder {
        name "d455_depth_camera"
        fieldOfView 1.518436
        width 1280
        height 720
        near 0.520000
        minRange 0.520000
        maxRange 4.000000
      }
```

Here's the result in RViz:

Screenshot from 2022-02-23 17-32-12

There are some projection issues. The RGBD image in RViz (greyscaled, top left) is identical to the depth image displayed by Webots itself. Therefore, I would argue that this issue happens already in Webots.

**Comment by astumpf** [comment]
> Still have some minor issues with the rangefinder, which I hope you can figure out quickly.
> ...

Solved the transformation issue. The problem was the missing exports of the optical frames in our URDF, so this mistake was on my side. But it requires still another bugfix (see #164).

However, warping remains a serious issue:

Screenshot from 2022-02-23 18-51-22

File: urdf2webots/parserURDF.py (hunk 1 above)
Comment (BenjaminHug8) [reply urdf2webots/parserURDF.py:904]: This part is still missing.

**Comment by BenjaminHug8** [comment]
I will take a look at this, may I ask you your minimal example URDF file so I can also compare with the same example as you have?

File: urdf2webots/parserURDF.py (hunk 1 above)
Comment (astumpf) [reply urdf2webots/parserURDF.py:904]: Yes, didn't had time to review these changes on my side.

**Comment by BenjaminHug8** [comment]
May I ask you your URDF used to generate this PROTO RangeFinder?

**Comment by astumpf** [comment]
> May I ask you your URDF used to generate this PROTO RangeFinder?

I will send you some files the next days.

**Comment by BenjaminHug8** [comment]
After some tests and according with the [documentation on the RangeFinder](https://www.cyberbotics.com/doc/reference/rangefinder), I think the `near` value is too high in your case. This value should be just big enough to exclude any geometry of the RangeFinder from the frustum.

Could you try to reduce it and report here the effect?

**Comment by astumpf** [comment]
> After some tests and according with the [documentation on the RangeFinder](https://www.cyberbotics.com/doc/reference/rangefinder), I think the `near` value is too high in your case. This value should be just big enough to exclude any geometry of the RangeFinder from the frustum.
> 
> Could you try to reduce it and report here the effect?

Set near (and in a second run minRange) to 0.1. Both tests results in following:

Screenshot from 2022-02-23 17-42-33

Curvature is removed, but plane is tilted. I'm just wondering that at max range definite values are reported. According to your [documentation](https://www.cyberbotics.com/doc/reference/rangefinder) +inf should be returned for those pixels, which also matches the [ROS REP117](https://www.ros.org/reps/rep-0117.html). Therefore, we should not be even able to "see" the far plane.

**Comment by astumpf** [comment]
> After some tests and according with the [documentation on the RangeFinder](https://www.cyberbotics.com/doc/reference/rangefinder), I think the `near` value is too high in your case. This value should be just big enough to exclude any geometry of the RangeFinder from the frustum.
> 
> Could you try to reduce it and report here the effect?

Set near (and in a second run minRange) to 0.1. Both tests results in following:

Screenshot from 2022-02-24 15-17-09

Curvature is removed, but plane is tilted. I'm just wondering that at max range definite values are reported. According to your [documentation](https://www.cyberbotics.com/doc/reference/rangefinder) +inf should be returned for those pixels, which also matches the [ROS REP117](https://www.ros.org/reps/rep-0117.html). Therefore, we should not be even able to "see" the far plane.

**Comment by BenjaminHug8** [comment]
> Curvature is removed,

Nice!

> but plane is tilted.

I will investigate that.

> `+inf` should be returned for those pixels

That's something we recently fixed. Are you using Webots R2022a or a nightly build?

**Comment by astumpf** [comment]
> > Curvature is removed,
> 
> Nice!
> 
> > but plane is tilted.
> 
> I will investigate that.
> 
> > `+inf` should be returned for those pixels
> 
> That's something we recently fixed. Are you using Webots R2022a or a nightly build?

Webots R2022a

**Comment by BenjaminHug8** [comment]
> Webots R2022a

Ok then trying one of the [latest nightly build](https://github.com/cyberbotics/webots/releases) should resolve the max range values.

I will get back to you once I have understand the tilt of the plane.

**Comment by BenjaminHug8** [comment]
I made a quick test with the rangeFinder and the result shows a bending for horizontal planes (see underneath). It seems to be affected by the near plane and is probably due to some mapping of the frustum of the camera, I am currently working on it.



The range finder used is the one you gave but with a near plane of 0.1:
```
RangeFinder {
        name "d455_depth_camera"
        fieldOfView 1.518436
        width 1280
        height 720
        near 0.1
        minRange 0.1
        maxRange 4.000000
      }
```

But otherwise I am not able to reproduce the tilt of the plane.

**Comment by astumpf** [comment]
> I made a quick test with the rangeFinder and the result shows a bending for horizontal planes (see underneath, the wall on the right is not the max range but a physical wall). It seems to be affected by the near plane and is probably due to some mapping of the frustum of the camera, I am currently working on it.
> 
> Screenshot from 2022-02-28 11-48-10
> 
> The range finder used is the one you gave but with a near plane of 0.1:
> 
> ```
> RangeFinder {
>         name "d455_depth_camera"
>         fieldOfView 1.518436
>         width 1280
>         height 720
>         near 0.1
>         minRange 0.1
>         maxRange 4.000000
>       }
> ```
> 
> But otherwise I am not able to reproduce the tilt of the plane.

It seems your model has also some tilted coordinate systems. Are you sure you are not rendering the depth image from one of these that could accidently compensate the tilt of the plane?
However, I've send you our proto file a few days ago.

**Comment by BenjaminHug8** [comment]
> It seems your model has also some tilted coordinate systems. Are you sure you are not rendering the depth image from one of these that could accidently compensate the tilt of the plane?

I can check this.

> However, I've send you our proto file a few days ago.

I ll assume that I just have to replace the lidar by a rangefinder in the PROTO file you sent.

**Comment by astumpf** [comment]
> I ll assume that I just have to replace the lidar by a rangefinder in the PROTO file you sent.

The proto is ready to run out of the box :-). No changes required.

**Comment by BenjaminHug8** [comment]
> The proto is ready to run out of the box :-). No changes required.

I am a bit confused, we are talking about rangeFinder in this PR and I received a Lidar.box.proto with a lidar device in it.

Were you talking about a lidar all this time or did I not have the correct file?
(The visual of the PROTO is the same, but I just surprised to found a lidar and not a rangeFinder, I will test with the lidar until you see this message).

**Comment by astumpf** [comment]
> > The proto is ready to run out of the box :-). No changes required.
> 
> I am a bit confused, we are talking about rangeFinder in this PR and I received a Lidar.box.proto with a lidar device in it.
> 
> Were you talking about a lidar all this time or did I not have the correct file? (The visual of the PROTO is the same, but I just surprised to found a lidar and not a rangeFinder, I will test with the lidar until you see this message).

I'm sorry for any confusion. This proto is indeed the "Lidarbox" model I'm using for the tests above. It includes both, Lidar and Rangefinder. Lidar works already fine for us.

**Comment by BenjaminHug8** [comment]
> I'm sorry for any confusion. This proto is indeed the "Lidarbox" model I'm using for the tests above. It includes both, Lidar and Rangefinder. Lidar works already fine for us.

Ah ok, I understand better now. I made a quick test with the rangeFinder and I do not see a tilt of the plane.

However if I rotate the device in Webots it appears that the visualization in RViz is not simply rotated, e.g. if I have a horizontal plane with a vertical wall, after a rotation in of the device in Webots I can see (of course) a rotation in Rviz but the plane and the wall are no longer perpendicular and I almost get a horizontal plane tilted but the wall is still vertical.

So I am wondering if you publish your tf tree from your URDF or if you let the `webots_ros2_driver` do it for you ([here is an example on how to set this option](https://github.com/cyberbotics/webots_ros2/blob/master/webots_ros2_tiago/launch/robot_launch.py#L77))?
If you use your own URDF I suspect the converter tool to maybe not correctly handle rotations.

**Comment by BenjaminHug8** [comment]
> I made a quick test with the rangeFinder and the result shows a bending for horizontal planes (see underneath, the wall on the right is not the max range but a physical wall). It seems to be affected by the near plane and is probably due to some mapping of the frustum of the camera, I am currently working on it.
> 
> Screenshot from 2022-02-28 11-48-10

This effect has been fixed in https://github.com/cyberbotics/webots/pull/4309.

**Comment by astumpf** [comment]
To give this topic new input. I'm currently experimenting with the latest changes introduced with the 2022b release.

This issue here unfortunately remains. Here is a new example when using a range finder:

```
RangeFinder {
  name "vision_d455_depth_camera"
  fieldOfView 1.518436
  width 1280
  height 720
  near 0.520000
  minRange 0.520000
  maxRange 4.000000
}
```

Webots world:
Screenshot from 2022-10-05 22-00-33

Produced depth image (converted to point cloud using [depth_image_proc](http://wiki.ros.org/depth_image_proc), but the rangefinder image is confirmed to contain the same data):
Screenshot from 2022-10-05 22-00-12

Now, a big concern which is hidden in this example: The color image is perfectly projected into the point cloud. This is an indication that the color image may be heavily warped as well. This would explain the strange behavior that AR marker detectors (e.g. [Aruco](https://docs.opencv.org/4.x/d5/dae/tutorial_aruco_detection.html)) have a hard time to detect anything in camera images rendered by Webots.

**Comment by astumpf** [comment]
Nevermind, this projection issue was on our side due to a wrongly published camera info. After sorting out this issue, everything works perfectly. This PR is ready to merge (finally!) :+1:

**Review by omichel — APPROVED** [review]
Looks good to me.
Thank you.

**Pull request merged**